### PR TITLE
added system logging utility

### PR DESCRIPTION
**Pull request by stevekuznetsov**
@deads2k @liggitt @smarterclayton FYI

### CODE REVIEW

File: hack/log.sh [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+# os::log::install_cleanup installs os::log::clean_up_logger as a trap on interrupts, termination, and exit
+#
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  None
+function os::log::install_cleanup() {
+    trap "os::log::clean_up_logger; $(trap -p SIGINT)" SIGINT
```
Comment (stevekuznetsov) [inline hack/log.sh]: will look upstream to determine best way to prefix traps

File: hack/log.sh [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+#  - SAR_LOGFILE
+# Arguments:
+#  None
+# Returns:  
+#  None
+function os::log::clean_up_logger() {
+    local return_code
+    return_code=$?
+
+    if jobs -pr | grep -q "${LOGGER_PID}"; then
```
Comment (stevekuznetsov) [inline hack/log.sh]: unclear if this block is necessary but it's here so we don't have runaway logger processes

File: hack/log.sh [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    for (( i = 0; i < "${#log_subset_flags[@]}"; i++ )); do
+        log_subset_file="${LOG_DIR}/${log_subset_names[$i]}.txt"
+        # use sadf utility to extract data into easily-parseable format
+        sadf -d "${SAR_LOGFILE}" -- ${log_subset_flags[$i]} > "${log_subset_file}"
+
+        local ignored_columns="hostname,interval,"
+
+        # special cases for special output from SAR, because the tool often gives us columns full of baloney
+        if [[ "${log_subset_names[$i]}" == "cpu" ]]; then
+            ignored_columns="${ignored_columns}CPU,"
```
Comment (stevekuznetsov) [inline hack/log.sh]: when asked for all CPU data, `sar` fills this column with `-1` ...

File: hack/log.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/log.sh]: *even though* I am doing this entire block on this `if` statement, which checks if the PID is in the list of running child processes of this PID, I *still* sometimes get:
```
/home/stevekuznetsov/go/src/github.com/openshift/origin/hack/log.sh: line 36: kill: (7094) - No such process
```

I have no idea how that is possible.

Comment (deads2k) [inline hack/log.sh]: > I have no idea how that is possible.

Cleanup happened async to it.  I don't think we care, so `set +e` or ` || true` would be appropriate.

File: hack/log.sh [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    local log_subset_file
+    local i
+    for (( i = 0; i < "${#log_subset_flags[@]}"; i++ )); do
+        log_subset_file="${LOG_DIR}/${log_subset_names[$i]}.txt"
+        # use sadf utility to extract data into easily-parseable format
+        sadf -d "${SAR_LOGFILE}" -- ${log_subset_flags[$i]} > "${log_subset_file}"
+
+        local ignored_columns="hostname,interval,"
+
+        # special cases for special output from SAR, because the tool often gives us columns full of baloney
```
Comment (deads2k) [inline hack/log.sh]: bologna

**Comment by stevekuznetsov** [comment]
[test] me

File: hack/log.sh (hunk 4 above)
Comment (stevekuznetsov) [inline hack/log.sh]: https://www.google.com/search?q=definebaloney&ie=utf-8&oe=utf-8#q=define+baloney

File: hack/log.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/log.sh]: This doesn't seem to trigger `set -o errexit` amazingly, either.

File: hack/log.sh (hunk 1 above)
Comment (stevekuznetsov) [inline hack/log.sh]: upstream only does append for traps. if we want this we need to be much more diligent about how we exit traps.

I'll address this in another pull.

**Comment by stevekuznetsov** [comment]
added logging for `test-cmd` because that's going to flake forever before we get to `test-integration`... re[test]

**Comment by stevekuznetsov** [comment]
@deads2k artifacts seem to have been created for `test-cmd` but not `test-integration`. No `test-integration` folder is downloaded with the artifacts. We need to either update the download script or deconflict the names if it is aliasing LOG_DIR to something else. Plots seen here: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8509/artifact/origin/artifacts/cmd/logs/ @liggitt  also PTAL what else do we want plotted?

**Comment by deads2k** [comment]
@stevekuznetsov how about e2e?

**Comment by deads2k** [comment]
@stevekuznetsov do you know where that download script is?  I'd really like to see it.

Also, how much tidying do you want to do before merge.  This is a great first step even if we start gathering more info as we move along.

File: hack/test-cmd.sh [hunk 5]
```diff
@@ -72,6 +73,9 @@ reset_tmp_dir
 
 echo "Logging to ${LOG_DIR}..."
 
+os::log::install_system_logger # after compiulation
```
Comment (liggitt) [inline hack/test-cmd.sh]: comments above lines, and without typos :)

**Comment by deads2k** [comment]
I think I'd like to see:
 1. What do timestamps on the x-axis look like?
 1. Timestamps for test-cmd.sh, test-end-to-end, and test-integration.sh


There was a spike in sys, steal, and reads all at the same time.  I'd like to know what that's about and I realized I couldn't see what it was doing then.

Also, do you have any ideas for automatically comparing runs to highlight outliers?  (not in this pull obviously)

**Comment by stevekuznetsov** [comment]
@deads2k the timestamp is at the bottom of all of the stacked plots

I had an idea to work timestamps into `os::cmd`, and write a parser that will query the `sar` database for data during the test execution, which we can then embed as metadata on jUnit XML

File: hack/test-integration.sh [hunk 6]
```diff
@@ -20,10 +21,7 @@ export ETCD_HOST=${ETCD_HOST:-127.0.0.1}
 export ETCD_PORT=${ETCD_PORT:-44001}
 export ETCD_PEER_PORT=${ETCD_PEER_PORT:-47001}
 
-
-set +e
```
Comment (stevekuznetsov) [inline hack/test-integration.sh]: every time I can remove a set of these bad boys I call it a win

**Comment by stevekuznetsov** [comment]
@deads2k download script is at  https://github.com/openshift/vagrant-openshift/blob/master/lib/vagrant-openshift/command/download_artifacts_origin.rb

**Comment by stevekuznetsov** [comment]
@deads2k left to do:
 1. make sure we download artifacts from `test-integration` https://github.com/openshift/vagrant-openshift/pull/383
 2. add this to all test scripts

**Comment by stevekuznetsov** [comment]
Got bitten by bad `trap` conventions again, so end-to-end didn't get a chance to plot. I'm thinking `test-integration` will run into similar issues. I may have to fix these trap issues first. We can merge this, but we won't see the plots until I fix the traps. The actual `sar` log should exist either way.

**Comment by deads2k** [comment]
> Got bitten by bad trap conventions again, so end-to-end didn't get a chance to plot. I'm thinking test-integration will run into similar issues. I may have to fix these trap issues first. We can merge this, but we won't see the plots until I fix the traps. The actual sar log should exist either way.

Go ahead and get the traps fixed.

**Comment by stevekuznetsov** [comment]
@deads2k trap PR here: https://github.com/openshift/origin/pull/6657

**Comment by stevekuznetsov** [comment]
We  can see ~100MB paged out to disk [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8557/artifact/origin/artifacts/integration/logs/paging.pdf) during the start of integration tests.

**Comment by deads2k** [comment]
@liggitt test-integration.sh is not CPU bound

**Comment by deads2k** [comment]
That test-integration behavior is neat to look at.  It might line up with writing the wal segment.  After this merges, I'll add some additional logging.

**Comment by stevekuznetsov** [comment]
@deads2k This looks reasonably ready... can't make any promises that cleanup will actually happen as advertised in every single case, but that's the best we can do for now.

File: hack/lib/log.sh [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+# Globals:
+#  - LOG_DIR
+#  - LOGGER_PID
+#  - SAR_LOGFILE
+# Arguments:
+#  None
+# Returns:  
+#  None
+function os::log::clean_up_logger() {
+    local return_code
```
Comment (deads2k) [inline hack/lib/log.sh]: call the directive to ignore errors.

File: hack/lib/log.sh [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+# Arguments:
+#  None
+# Returns:
+#  - export LOGGER_PID
+#  - export SAR_LOGFILE
+function os::log::install_system_logger() {
+    readonly SAR_LOGFILE="${LOG_DIR}/sar.log"
+    export SAR_LOGFILE
+
+    ( os::log::internal::run_system_logger "${SAR_LOGFILE}" "${LOG_DIR}/sar_stderr.log" ) &
```
Comment (deads2k) [inline hack/lib/log.sh]: check for existence of `sar` before trying to use it.  Don't fail on not having it, but echo a message.

File: hack/lib/log.sh [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+#  - SAR_LOGFILE
+# Arguments:
+#  None
+# Returns:  
+#  None
+function os::log::clean_up_logger() {
+    local return_code
+    return_code=$?
+
+    if jobs -pr | grep -q "${LOGGER_PID}"; then
```
Comment (deads2k) [inline hack/lib/log.sh]: If `sar` isn't present, this should be a no-op

File: hack/lib/log.sh [hunk 10]
```diff
@@ -0,0 +1,216 @@
+#!/bin/bash
```
Comment (deads2k) [inline hack/lib/log.sh]: in some future pull, this seems to be a reasonable spot for our existing log functions and green,red, etc.

File: hack/lib/log.sh [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+# os::log::install_system_logger installs the system logger and begins logging
+#
+# Globals:
+#  - LOG_DIR
+# Arguments:
+#  None
+# Returns:
+#  - export LOGGER_PID
+#  - export SAR_LOGFILE
+function os::log::install_system_logger() {
```
Comment (deads2k) [inline hack/lib/log.sh]: This isn't installing.  `start_resource_logging` would be better.

File: hack/lib/log.sh (hunk 10 above)
Comment (stevekuznetsov) [inline hack/lib/log.sh]: Yep

**Comment by stevekuznetsov** [comment]
@deads2k addressed comments

**Comment by deads2k** [comment]
lgtm, but I want to see the output pdfs before merging.

**Comment by deads2k** [comment]
[merge]

**Pull request merged**